### ISSUE TITLE
[HELP WANTED] API refactoring

### ISSUE THREAD

**Issue by alexandreteles** [revanced/revanced-releases-api#77]
The current API design is based on a polling system that checks GitHub's REST API for data and caches it to reduce the number of requests made by the manager and website to the GH backend. It works well: it's fast and capable of withstanding millions of daily requests. Despite that, changes must be implemented to better integrate the API into our plans for how the patcher and the manager work.

That being said, those changes are sometimes too significant to be decided on by a single person. That's why I'm opening this issue: I need help from the team and the community to decide on architectural changes to our API. But what do we need?

### 1. Implement a push system so our repositories can push data to the API instead of polling.

Our initial idea was to build a server-client architecture where our Actions would have to authenticate against the API to push data into our database. That system would work fine, but it introduces complications and security concerns that are unnecessary for such a scenario. My current suggestion is to implement webhooks for each endpoint and receive data from GitHub exclusively when there is a new release. Those webhooks can be secured using a shared secret available to the team in a PSONO instance.

### 2. The changelog endpoint needs refactoring

The current implementation of changelogs is completely broken and needs rewriting. The first issue is *how* changelogs are built from the data we receive from GitHub: we check for changes to a specific file between the two newest releases. There are no proper guards in case of no changes, and the information received always refers to the whole set of changes between those releases. The entire infrastructure behind this endpoint must change to solve this issue. It will still pull data from GH API, but *how* we will build tailored changelogs is still a question mark. If you have any ideas, please let me know.

### 3. Filtering and pagination

The `/patches` endpoint will grow. How much? That's impossible to predict. Therefore, requesting the whole dataset as a payload might become increasingly taxing. Implementing pagination using existing tools is broken on Python 3.11, and reverting the migration to the newest Python version would increase our technical debt considerably. Filtering patches by name, so only the data referring to a specific patch is returned, is a desirable feature, but that would also require rewriting the whole infrastructure behind this endpoint. GraphQL would solve most of those problems but would require changing the business logic behind our endpoints or writing a new endpoint to deal exclusively with GraphQL queries. What do you think?

### 4. Database backend

One of the main reasons why our API is so robust (despite being written in Python) is Redis. We use Redis in a very exotic way, saving our data to disk every so often and keeping memory and disk in sync. Redis is **fast** and moving away from it to a more *traditional* DBMS would sacrifice a lot of our ability to serve a high amount of requests. At the same time, the complexity of our data is increasing (with the planned CDN and announcement endpoints), and although Redis *can* be used to store semi-structured data, it wasn't designed to be used as a permanent key-value store. Of course, this issue could easily be solved using something like GaussDB for Redis, but commercial solutions are obviously out of the question. So, we have the following options to solve our issue:

a. Keep using Redis as a permanent key-value datastore, implementing appropriate guards against data loss if the database goes down during the sync operation;

b. Move the data store to Aerospike, ScyllaDB, or ArangoDB and keep Redis only as our cache/authentication layer.

I carefully considered MongoDB, which would be a suitable solution, but I'm not too fond of it (or Beanie). Feel free to change my mind.

### 5. Integrating new services

Our polling API is currently set up as an entirely different service. It's a distinct code base, another repository, a different Docker container, and a separate subdomain. Ideally, we should have all our APIs under the same domain, `api.revanced.app`, and routing would be done to redirect requests to their respective microservices. That is possible and doable, but it would also make our setup more complicated (you have to be conscious of possible routing in front of your API when writing its routers; you need to keep track of which routes go where on your proxy configuration, etc.)

I prefer microservices to have their isolated subdomains whenever possible, but I'm here to listen to new solutions.

### 6. What else?

If you have any other suggestions for the API, please let's consolidate them all in this issue until we devise a proper plan on how to progress.

Thank you!

**Comment by Sculas** [comment]
Well written issue, good job!

1. What is the problem with polling to begin with? A PAT can invoke tons of requests per hour, more than you would ever need. I don't think it's necessary to switch this approach for the foreseeable future.

2. Unless I misunderstood how the CL route functions, can't you just return the description of the latest (pre-)release?

3. I know this route is used in the website, but does it really have any use beside that? Can't you just 301 to a GitHub raw link for the `patches.json` file? This still allows you to change the URL on the fly if the repo would go down without the users having to update.

4. Hmm.. this is a good one. ScyllaDB is a good option, but quite complex to set-up in my experience and definitely has a steep learning curve. You could also look at something like CockroachDB, which is similar to PostgreSQL but in some points better (like scaling). Though, if you want to stay with Redis, there's RedisJSON too. However, if you're willing to allocate the time, I would definitely recommend giving ScyllaDB a go. It's very promising!

5. Microservices will cause complexity and server costs to increase, which is not a viable option in this case. You could look at the Polylith architecture, it's pretty nice. But there's nothing wrong with sticking with the monolith architecture as well, in my opinion.

That's all. I've typed this all on my phone so I wasn't able to provide links/references, but I hope some things I've said are useful to you.

**Comment by oSumAtrIX** [comment]
Thanks for considering our feedback. Let's go straight into it!

## 1. Polling

This is a tricky one right from the beginning. The API depends on the repositories, but the repositories do not depend on the API. 

### It goes:

```mermaid
graph TD;
A[API] -->|Depends on| B(Repository)
```

### Your idea looks like this:

```mermaid
graph TD;
A[API] -->|Depends on| B(Repository)
B -->|Aware of| A
```

Let's understand the problem with this. The new relation is created with this:

> receive data from GitHub exclusively

The API receives data from GitHub instead of polling, reducing less strain on the API.  The trade for this is a new relationship between the API and the repository. With the new approach, both the API and the repository must be aware of each other, which burdens deployment. Consider the following scenario. I set up the API as a forker and want to provide a repository from someone else. But because the relationship is bidirectional, it requires the repository owner to set up the repository in order for the API to work, which is not always possible, for example, when the fork is outside an organization, and thus no webhook can be set up for it, but instead, a custom workflow has to be provided. What if multiple APIs depend on the repository, and what if the repository is outside of GitHub?

## 2. Changelogs

Changelogs are programmatically generated by inference from commits following the semantic commits standards. Because there is no bi-directional relationship between the repositories and the API, the repositories are not aware of the API. The changelog can either be generated from the commits or from releases. Both assume the repositories follow the API's specification and thus would require awareness of the API. Otherwise, a change in the repository can break the API. To get rid of this relationship and retain a non-bi-directional relationship, the repositories must provide a common specification on how to open changelogs are generated. Due to the API depending on the repositories, the API can now follow the specifications on the changelogs. Another issue should be opened and discussed thoroughly at the organizational level for a specification to clarify needs.

## 3. Pagination

The current design of our clients leads to the wrong approach of needing a `/patches`  endpoint. To begin with, the API does not need to provide a `/patches` endpoint at all because clients are supposed to load patches from the `.jar` files themselves. Though not all clients can load the patches from the `.jar` files, and thus a question remains in this regard.

## 4. Database

```
/
```

## 5. Services

Semantically speaking, the microservices in question belong to the domain `revanced.app`. The current design uses sub-domains which complicate architecture on the side of the clients. Instead, complexity should be moved to the server by using microservices because of a 1-N relationship (1 server, n clients). This would allow simple integration of clients with the API without having to deal with complexity n times.

### Current design:

```mermaid
graph TD;
A[API] --> |Simple integration with| Microservice
1[Client 1] --> |Complex dependency on| A
2[Client 2] --> |Complex dependency on| A
3[Client 3] --> |Complex dependency on| A
```

### Proposed design:

```mermaid
graph TD;
A[API] --> |Complex integration with| Microservice
1[Client 1] --> |Simple dependency on| A
2[Client 2] --> |Simple dependency on| A
3[Client 3] --> |Simple dependency on| A
```

**Comment by alexandreteles** [comment]
> 1. What is the problem with polling to begin with? A PAT can invoke tons of requests per hour, more than you would ever need. I don't think it's necessary to switch this approach for the foreseeable future.

* The main issue is that in rare cases the results provided by the API might not reflect reality. For example, let's say a given client requests the `/tools` endpoint right after a new release was done on `revanced-patches` but before `revanced-integrations` released a new version. That would introduce an inconsistency between two tools that depend on each other. That is an architectural issue in regard to how patches and integrations work, but it also affects the API. In that case, the API would provide inconsistent data for the next 5 minutes and could break patching in the manager.

> 2. Unless I misunderstood how the CL route functions, can't you just return the description of the latest (pre-)release?

* This endpoint was made under request from the Manager Team, and the idea behind it was to provide changelogs tailored by patch. How is it done? By checking for changes on specific files and listing the changes on the last release that changed said file. I would like to have a proper changelog per patch, so I agree with @oSumAtrIX that this requires a deeper analysis before we can effectively decide on what to do.

> 3. I know this route is used in the website, but does it really have any use beside that? Can't you just 301 to a GitHub raw link for the `patches.json` file? This still allows you to change the URL on the fly if the repo would go down without the users having to update.

* I could, but then I would be unable to apply any transformations to the file or guarantee its validity, transferring that responsibility to the clients. Ideally, the API would also get that information from the JAR file, but that would be considerably slower (JVM boot times, etc.) and still require polling now and then to check for new versions of the patch bundle. Of course, that process would happen once every release, and then the data would be cached, but it's still a pain to interop with Java from Python (and no, I'm not running the API inside of GraalVM).

> 4. Hmm.. this is a good one. ScyllaDB is a good option, but quite complex to set-up in my experience and definitely has a steep learning curve. You could also look at something like CockroachDB, which is similar to PostgreSQL but in some points better (like scaling). Though, if you want to stay with Redis, there's RedisJSON too. However, if you're willing to allocate the time, I would definitely recommend giving ScyllaDB a go. It's very promising!

* We currently use RedisJSON and RedisSearch for the API. The kind of data we store goes well with a key-value store, and Redis is amazing for our use case. The only reason we would have issues is in the rare case of the Redis server going down during the sync-to-disk operation. ScyllaDB is a pain to set up, and the Python drivers are completely synchronous, but as it would work as a data store, the first call to the API being slower isn't a problem (queries are small anyway, and they will be cached). I also considered using RocksDB, and being honest with you, among the DBMS I mentioned, ArangoDB is probably my favorite. In retrospect, I think that we should seriously consider RocksDB as a viable data store for this project.

> 5. Microservices will cause complexity and server costs to increase, which is not a viable option in this case. You could look at the Polylith architecture, it's pretty nice. But there's nothing wrong with sticking with the monolith architecture as well, in my opinion.

* The project is currently very Pythonic, and moving towards a polylithic architecture would require a lot of effort on rewrites. I'm not entirely sure it would bring enough advantages to balance the technical debt that would accumulate. I'll consider that and look into other possibilities more carefully. The last thing I want is for the API to become a huge monolith because I'm currently the sole maintainer in this project, and my brain has a serious skill issue remembering things inside of a monolithic project (especially if it's a big one).

**Comment by alexandreteles** [comment]
> Consider the following scenario. I set up the API as a forker and want to provide a repository from someone else. But because the relationship is bidirectional, it requires the repository owner to set up the repository in order for the API to work, which is not always possible, for example, when the fork is outside an organization, and thus no webhook can be set up for it, but instead, a custom workflow has to be provided. 

In this case, a configuration setting can be provided so whoever is hosting their own API instance can toggle between polling and webhooks. This would allow the flexibility of receiving updates when they happen instead of being up to five minutes late and also reduce the risk of inconsistencies in the `/tools` endpoint.

> What if multiple APIs depend on the repository, and what if the repository is outside of GitHub?

The first part is solved by providing a setting that allows for polling. The second part is a bit trickier. Every source code hosting service has a different API. With webhooks, you can call the API anytime during your CI/CD task or use the interface for webhooks provided by your service provider. The dependence on GH API is a byproduct of where *we* host our code, but writing new interfaces for different providers is fairly simple considering our current architecture. The only reason it doesn't exist is that I never received a request for it.

> To get rid of this relationship and retain a non-bi-directional relationship, the repositories must provide a common specification on how to open changelogs are generated. Due to the API depending on the repositories, the API can now follow the specifications on the changelogs. Another issue should be opened and discussed thoroughly at the organizational level for a specification to clarify needs.

Could you please move this discussion upstream so we can look into what can be done concerning a common specification?

> The current design of our clients leads to the wrong approach of needing a `/patches` endpoint. To begin with, the API does not need to provide a `/patches` endpoint at all because clients are supposed to load patches from the `.jar` files themselves. Though not all clients can load the patches from the `.jar` files, and thus a question remains in this regard.

Regarding the current file size issue, I recommend that we offer a `?format=msgpack` query for the `/patches` endpoint that returns a *much* smaller Brotli compressed msgpack file. I'll make tests on loading that information directly from the JAR file but I don't expect it to be either computationally cheap or especially fast.

> Semantically speaking, the microservices in question belong to the domain `revanced.app`. The current design uses sub-domains which complicate architecture on the side of the clients. Instead, complexity should be moved to the server by using microservices because of a 1-N relationship (1 server, n clients). This would allow simple integration of clients with the API without having to deal with complexity n times.

That is a good point, but I'm still not 100% convinced. I'll look into that issue again later today, but having to make the APIs aware of the proxy in front of them is work I'm not looking forward to, and I might adopt a big monolithic solution to save time. Microservices can be put behind routes, obviously. Still, as the number of microservices grows and they move away from having their root subdomains, I'll have to make the application aware of *where* they are being served (behind which route). That's not particularly difficult to implement; on the contrary, it's just annoying.

> Currently, the API automatically polls and provides the latest version of the repository releases. It might be desired to control this and pin a specific release version.

The API lacks a concept of release that would allow pinning. I'll have to think about implementing that solution properly, but it'll probably be a `config.toml` variable.

**Comment by oSumAtrIX** [comment]
> In this case, a configuration setting can be provided so whoever is hosting their own API instance can toggle between polling and webhooks. This would allow the flexibility of receiving updates when they happen instead of being up to five minutes late and also reduce the risk of inconsistencies in the /tools endpoint.

This would work. However, the motivation for making changes to a repository to push data would be out of awareness of the API, creating a bi-directional relationship. The API would depend on the awareness of said repository, which complicates things. This question involves the trade-off between making the API dependent on the repositories configuration and making the repositories aware of the API and having a non-bi-directional relationship so that the repository does not need to consider the existence of the API but with the trade of the API having to poll. 

If we decide not to use polling, the API repository should provide a specification on making a repository aware of it so all of our necessary repositories can adapt them according to the spec.

> Could you please move this discussion upstream so we can look into what can be done concerning a common specification?

A [discussion](https://github.com/orgs/revanced/discussions/891) has been reserved for this.

> Regarding the current file size issue, I recommend that we offer a ?format=msgpack query for the /patches endpoint

Is the `/patches` endpoint even necessary? I think it does not belong to `revanced-releases-api` to begin with. The reason is that this specific API should be treated as a generic "GitHub repository releases proxy". This means it should provide a transformable layer for GitHub releases. On the other hand, the `/patches` endpoint would rather belong in a different API, such as a `revanced-api`:

```mermaid
graph TD
    Client -->|Served by| API[api.revanced.app]
    API --> |Extended by| R[api.revanced.app/releases]
    R --> |Provides| T[GitHub releases transformer]
    API --> |Provides| P[General purpose API] 
    P --> |Includes| Patches[Patches endpoint]
  ```

> but having to make the APIs aware of the proxy in front of them is work I'm not looking forward to

I don't think the microservices must be aware of the proxy in front of them. After all, the proxy (`api.revanced.app`) would simply upstream requests to the microservice; for example, the GitHub releases proxy API (`revanced-releases-api`).  

>  I'll have to make the application aware of where they are being served (behind which route)

Can you clarify which application you are referring to? Does the app behind the proxy have to be aware of the route or the proxy itself? I don't think the latter is an issue.

> The API lacks a concept of release that would allow pinning. I'll have to think about implementing that solution properly, but it'll probably be a `config.toml` variable.

The original issue was that the API lacks flexibility. Usually, this isn't a problem, but what if it is necessary to avoid a specific release?
